phabfive/maniphest.py

```diff
@@ -49,115 +51,247 @@ def create_from_config(self, config_file, dry_run=False):
+                for task in child_tasks:
+                    processed_child_tasks.append(
+                        recurse_build_transactions(task)
+                    )
+            else:
+                processed_child_tasks = None
+
+            output["tasks"] = processed_child_tasks
+
+            return output
+
+        def recurse_commit_transactions(task_config, parent_task_config):
+            """
+            This recurse functions purpose is to iterate over all tickets, commit them to phabricator
+            and link them to eachother via the ticket hiearchy or explicit parent/subtask links.
+
+            task_config is the current task to create and the parent_task_config is if we have a tree
+            of tickets defined in our config file.
+            """
+            log.debug("\n -- Commiting task")
+            log.debug(json.dumps(task_config,indent=2))
+            log.debug(" ** parent block")
+            log.debug(json.dumps(parent_task_config, indent=2))
+
+            transactions_to_commit = task_config.get("transactions", [])
+
+            if dry_run:
+                log.critical("Running with --dry-run, tickets !!WILL NOT BE!! commited to phabricator")
+            else:
+                if transactions_to_commit:
+                    # Parent ticket based on the task hiearchy defined in the config file we parsed is different
+                    # from the explicit "ticket parent" that can be defined 
+                    if parent_task_config and "phid" in parent_task_config:
+                        transactions_to_commit.append({
+                            "type": "parents.add",
+                            "value": [parent_task_config["phid"]],
+                        })
+
+                    log.debug(" -- transactions to commit")
+                    log.debug(transactions_to_commit)
+
+                    if dry_run:
+                        log.critical("Running with --dry-run, tickets !!WILL NOT BE!! commited to phabricator")
+                    else:
+                        result = self.phab.maniphest.edit(
+                            transactions=transactions_to_commit,
+                        )
+
+                    # Store the newly created ticket ID in the data structure so child tickets can look it up
+                    task_config["phid"] = str(result["object"]["phid"])
+                else:
+                    # log.warning(f"No transactions found for ")
+                    log.warning("No transactions to commit here, either a bug or root object that can't be transacted")
+
+            child_tasks = task_config.get("tasks", None)
+            if child_tasks:
+                for child_task in child_tasks:
+                    recurse_commit_transactions(child_task, task_config)
+
+        # Main task recursion logic
+        if "tasks" not in root_data:
+            log.critical(f"Config file must contain keyword tasks in the root")
+            return 1
```

Comment:
Return numeric `1`? If this is to indicate an error, then an exception should be used? If it is a common case that nothing is processed then perhaps return an empty list instead?

Reply:
So there is a bigger design philosophy question here that we have discussed internally several times. Either you design libraries with no error returning, or you return status codes like this example, or you raise exception that someone else has to handle for you. Returning an empty object in this case can be very misleading because then you would have to always interprete that empty list is an error and not a valid result output from the function call.

I modified it to an exception as wen can try that and see how that works out